drivers/eeprom Add support for eeprom simulator

Laczen opened the pull request:
Add support for a eeprom simulator. The PR limits the addition to
qemu_x86 but it can easily be added to other devices by defining the
eeprom simulator in the dts and setting 'CONFIG_EEPROM_SIMULATOR=y'

Signed-off-by: Laczen JMS <laczenjms@gmail.com>

Laczen committed:
drivers: eeprom: Add support for eeprom simulator

Add support for a eeprom simulator. The PR limits the addition to
qemu_x86 but it can easily be added to other devices by defining the
eeprom simulator in the dts and setting 'CONFIG_EEPROM_SIMULATOR=y'

Signed-off-by: Laczen JMS <laczenjms@gmail.com>
Signed-off-by: Anas Nashif <anas.nashif@intel.com>

henrikbrixandersen commented:
@Laczen Looks good, thank you. Do we need locking to protect against multiple reads/writes?

henrikbrixandersen commented:
@Laczen Great. Please squash the two commits into one and use `drivers: eeprom: ` instead of `drivers/eeprom: ` in the commit log messages to follow the standard commit message format.

Laczen commented:
> @Laczen Great. Please squash the two commits into one and use `drivers: eeprom: ` instead of `drivers/eeprom: ` in the commit log messages to follow the standard commit message format.

@brix, I like it when people can cherrypick a PR, that's why I didn't squash.

Regarding the commit log messages: I have created quite some PR's for the subsys dir, these always fail when using subsys: xxx, so I am always using subsys/xxx or drivers/yyy. It's good to have standards if they work...

henrikbrixandersen commented:
> @brix, I like it when people can cherrypick a PR, that's why I didn't squash.

Not sure I understand why squashing two commits would prevent cherry picking?
 
> Regarding the commit log messages: I have created quite some PR's for the subsys dir, these always fail when using subsys: xxx, so I am always using subsys/xxx or drivers/yyy. It's good to have standards if they work...

Commits to the `subsys/xxx` folder should follow the format `xxx:` not `subsys: xxx:`. Not sure what you mean by having standards that work (and how it relates to this PR).

vanwinkeljan commented on drivers/eeprom/eeprom_simulator.c:
```diff
@@ -0,0 +1,209 @@
+/*
+ * Copyright (c) 2018 Nordic Semiconductor ASA
```
Wrong copyrights

vanwinkeljan reviewed (commented):
Wondering if you should not merge the native_posix driver in to this driver.
In the end the native_posix eeprom driver just memmaps a file so effort should be minimal.

Offtopic: I have been thinking to do the same for the flash variant, merge native_posix into flash sim driver, ~but never found the time for it~ (see: PR #20553). The main reason why both drivers exists is that the initial simulated flash driver had a size limitation that was not acceptable for native_posix and both PRs where under review at the same time.

Laczen commented:
> Wondering if you should not merge the native_posix driver in to this driver.
> In the end the native_posix eeprom driver just memmaps a file so effort should be minimal.
> Offtopic: I have been thinking to do the same for the flash variant, merge native_posix into flash sim driver, but never found the time for it (see: PR #20553). The main reason why both drivers exists is that the initial simulated flash driver had a size limitation that was not acceptable for native_posix and both PRs where under review at the same time.

That is a great idea, I will rework like 20553 if @brix agrees (removal of eeprom,native_posix).

henrikbrixandersen commented:
> That is a great idea, I will rework like 20553 if @brix agrees (removal of eeprom,native_posix).

Yes, sounds good. Please do.

vanwinkeljan commented:
@henrikbrixandersen My apologies for not bringing this topic up during review of the native_posix driver. Completely forgot about my idea to merge both flash drivers.

> @Laczen Looks good, thank you. Do we need locking to protect against multiple reads/writes?

Note that simulator can also be used under QEMU in this case there could be concurrently running threads.

Laczen committed:
drivers: eeprom: Unified simulator and native_posix

EEPROM simulator and native_posix have been unified to one solution,
the old eeprom,native_posix is removed.

Signed-off-by: Laczen JMS <laczenjms@gmail.com>

vanwinkeljan commented on drivers/eeprom/eeprom_simulator.c:
```diff
@@ -193,17 +212,99 @@ static const struct eeprom_sim_config eeprom_sim_config_0 = {
 	STATS_INIT_AND_REG(eeprom_sim_thresholds, STATS_SIZE_32,
 			   "eeprom_sim_thresholds");
-	memset(mock_eeprom, 0xFF, ARRAY_SIZE(mock_eeprom));
+	eeprom_mock_init(dev);
```
Better return the value of the function, just in case the posix arch version of`eeprom_mock_init ` returns an error code.

aescolar commented on drivers/eeprom/eeprom_simulator.c:
```diff
@@ -0,0 +1,309 @@
+/*
+ * Copyright (c) 2019 Laczen
```
Note that assigning copyrights would be done to either real people or companies. So if you want to write here a copyright notice for you, you would need to write your name, not your GH account name.

Laczen replied:
Laczen is my company name, and also my github account name :smile:

aescolar replied:
aa ,ok. If it is an actual company then nothing :)

aescolar commented on drivers/eeprom/eeprom_simulator.c:
```diff
@@ -0,0 +1,309 @@
+
+	eeprom_fd = open(eeprom_file_path, O_RDWR | O_CREAT, (mode_t)0600);
+	if (eeprom_fd == -1) {
+		posix_print_warning("Failed to open eeprom device file "
```
Same comment as here: https://github.com/zephyrproject-rtos/zephyr/pull/20553/files/c631a4defa8ef73f9f130c0c2ab2c79b917b8556#diff-19518c7a115928db382df066c14dbe29

Laczen replied:
@aescolar, do you mean use `posix_print_error()` or something else ?

aescolar replied:
yep

Laczen replied:
I am not sure if we should exit, just return an error that init failed.

aescolar replied:
I guess it depends on what will happen later if the app does not handle it and tries to access the  eprom anyhow (segfault?). I have no strong opinion anyhow here.

Laczen replied:
I would rather keep it as is, this allows you to check that the app handles it, which it should.

aescolar reviewed (commented):
Just 2 relatively minor issues

henrikbrixandersen commented on drivers/eeprom/eeprom_simulator.c:
```diff
@@ -0,0 +1,309 @@
+		}
+
+		*(EEPROM(offset + i)) = *((u8_t *)data + i);
+	}
```
What is the purpose of iterating through all the bytes? Limiting the number of bytes written overall based on `max_len`? If so, wouldn't it be sufficient to cap `len` to `max_len` and use `memcpy()` or am I missing something?

Laczen replied:
@brix, this allows you to simulate write errors, anything that is beyond the max_len is ignored. An example of how this can be used can be found in tests/subsys/fs/nvs.

henrikbrixandersen replied:
@Laczen I get that much. But couldn't you just do something like `MAX(len, max_len)` and do a `memcpy()`?

Laczen replied:
@brix, updated. Made me realize that there was something wrong with the lock, corrected.

henrikbrixandersen commented:
@Laczen Please rebase on master to resolve the merge conflict.

vanwinkeljan commented:
@Laczen you will need to rebase on top of the master branch instead of merging it

Laczen commented:
> @Laczen you will need to rebase on top of the master branch instead of merging it

I tried to resolve the conflict directly in github, seems this was not the correct way. Will do the rebase if I find some spare time.

aescolar commented:
Just labeling DNM to be sure nobody accidentally merges that merge commit. (Anybody,) please remove the label when the branch is rebased.

carlescufi commented:
@Laczen can you please rebase so we can merge?
